README.md
# Gamify

[Preview Gamify](https://gamify-dacade.herokuapp.com)

[Preview emmauelJet dacade Review](https://emmanueljet.github.io/gamify)

This project was bootstrapped with [Create React App](https://github.com/facebook/create-react-app) while utilizing [Celo Blockchain](http://celo.org) technology for the [dacade Celo Development 101 course](https://dacade.org/communities/celo-development-101/submissions/7463cb3f-d328-4708-800a-dc9fef03f453).

## Available Scripts

In the project directory, you can run:

### `npm start`

Runs the app in the development mode.\
Open [http://localhost:3000](http://localhost:3000) to view it in the browser.

The page will reload if you make edits.\
You will also see any lint errors in the console.

### `npm test`

Launches the test runner in the interactive watch mode.\
See the section about [running tests](https://facebook.github.io/create-react-app/docs/running-tests) for more information.

### `npm run build`

Builds the app for production to the `build` folder.\
It correctly bundles React in production mode and optimizes the build for the best performance.

The build is minified and the filenames include the hashes.\
Your app is ready to be deployed!

See the section about [deployment](https://facebook.github.io/create-react-app/docs/deployment) for more information.

### `npm eject`

**Note: this is a one-way operation. Once you `eject`, you can’t go back!**

If you aren’t satisfied with the build tool and configuration choices, you can `eject` at any time. This command will remove the single build dependency from your project.

Instead, it will copy all the configuration files and the transitive dependencies (webpack, Babel, ESLint, etc) right into your project so you have full control over them. All of the commands except `eject` will still work, but they will point to the copied scripts so you can tweak them. At this point you’re on your own.

You don’t have to ever use `eject`. The curated feature set is suitable for small and middle deployments, and you shouldn’t feel obligated to use this feature. However we understand that this tool wouldn’t be useful if you couldn’t customize it when you are ready for it.

## Learn More

You can learn more in the [Create React App documentation](https://facebook.github.io/create-react-app/docs/getting-started).

To learn React, check out the [React documentation](https://reactjs.org/).

### Code Splitting

This section has moved here: [https://facebook.github.io/create-react-app/docs/code-splitting](https://facebook.github.io/create-react-app/docs/code-splitting)

### Analyzing the Bundle Size

This section has moved here: [https://facebook.github.io/create-react-app/docs/analyzing-the-bundle-size](https://facebook.github.io/create-react-app/docs/analyzing-the-bundle-size)

### Making a Progressive Web App

This section has moved here: [https://facebook.github.io/create-react-app/docs/making-a-progressive-web-app](https://facebook.github.io/create-react-app/docs/making-a-progressive-web-app)

### Advanced Configuration

This section has moved here: [https://facebook.github.io/create-react-app/docs/advanced-configuration](https://facebook.github.io/create-react-app/docs/advanced-configuration)

### Deployment

This section has moved here: [https://facebook.github.io/create-react-app/docs/deployment](https://facebook.github.io/create-react-app/docs/deployment)

### `npm run build` fails to minify

This section has moved here: [https://facebook.github.io/create-react-app/docs/troubleshooting#npm-run-build-fails-to-minify](https://facebook.github.io/create-react-app/docs/troubleshooting#npm-run-build-fails-to-minify)


CHANGELOG.md
# Gamify ChangeLog

Change Log file for the dacade [Celo Development 101](https://dacade.org/communities/celo-development-101) dApp submission by [GoodnessEzeokafor](https://dacade.org/communities/celo-development-101/submissions/7463cb3f-d328-4708-800a-dc9fef03f453).

**```Keywords```**

* `Added` for new features.
* `Changed` for changes in existing functionality.
* `Deprecated` for soon-to-be removed features.
* `Removed` for now removed features.
* `Fixed` for any bug fixes.
* `Security` in case of vulnerabilities.

## [emmanuelJet](https://github.com/emmanuelJet) - 2021-07-31

Great submission and nice implmentation of IPFS. **Contract Review:** I optimized the contract code by changing the way a game image is saved using IPFS, changed the code structure of some functions (buyGame, updateGamePrice, updateGameStatus), added a Sale functionality to record all sales details (Game ID, Buyer, Seller, Amount and Timestamp) then made the getSale() restricted and accessible by the contract admin (onlyContractAdmin modifier). **DApp Review:** I optimized the dApp to work with the updated smart contract codes and added a GitHub Workflow script to automatically build and deploy the dApp using GitHub Pages. **Design Review:** I did a rebrand of *`Celo Games`* to *`Gamify`* and made then fixed the favicon image bug. Also, I displayed the user CELO Balance, restructured the display of the `Sell your Games` form, updated the Footer Copyright position, and fixed the download game window bug after a successfull payment has been made.

### Added

* build_and_deploy.yml file
* .editorconfig file
* CHANGELOG.md file

### Changed

* Contracts with its JSON abi files
* package.json file
* README.md file
* index.html file
* App.js, Body.js, Footer.js, and Header.js files

### Removed

* yarn.lock file


src/App.js
import React, { useEffect, useState } from "react";
import { newKitFromWeb3 } from "@celo/contractkit";
import "@celo-tools/use-contractkit/lib/styles.css";
import BigNumber from "bignumber.js";
import gamifyABI from "./contracts/abi/gamify.abi.json";
import erc20Abi from "./contracts/abi/erc20.abi.json";
import { create, urlSource } from "ipfs-http-client";

import Web3 from "@celo/contractkit/node_modules/web3";
import Header from "./components/Header";
import Body from "./components/Body";
import Footer from "./components/Footer";

const ContractAddress = "0x874069Fa1Eb16D44d622F2e0Ca25eeA172369bC1";
const gamifyaddress = "0xf28d9714824796a76283CbDB53B96B3011c09BA7";

// import ipfs

/* Create an instance of the IPFS client */

const ipfsClient = create("https://ipfs.infura.io:5001/api/v0");

export default function App() {
  const [loading, setloading] = useState(false);

  const [celoBalance, setCeloBalance] = useState(0);
  const [cUSDBalance, setcUSDBalance] = useState(0);
  const [games, setGames] = useState([]);

  const [kit, setKit] = useState(null);
  const [address, setAddress] = useState(null);
  const [contract, setcontract] = useState(null);
  
  // the file url
  const [fileUrl, updateFileUrl] = useState(``);

  const ERC20_DECIMALS = 18;

  useEffect(() => {
    // connect the users wallet
    connectCeloWallet();
  }, []);

  const connectCeloWallet = async () => {
    if (window.celo) {
      // notification("⚠️ Please approve this DApp to use it.")
      try {
        setloading(true)
        await window.celo.enable();
        // notificationOff()
        const web3 = new Web3(window.celo);
        let kit = newKitFromWeb3(web3);

        const accounts = await kit.web3.eth.getAccounts();
        const user_address = accounts[0];
        kit.defaultAccount = user_address;

        const contract = new kit.web3.eth.Contract(gamifyABI, gamifyaddress);

        setKit(kit);
        setAddress(user_address);
        setcontract(contract);
      } catch (error) {
        console.log({ error });
        setloading(false)
      }
    } else {
      console.log("please install the extension");
      alert("⚠️ Please install the CeloExtensionWallet.")
    }
  };

  useEffect(() => {
    if (kit && address) {
      return getBalance();
    } else {
      console.log("no kit or address");
    }
  }, [kit, address]);

  const getBalance = async () => {
    setloading(true)
    const balance = await kit.getTotalBalance(address);
    const celoBalance = balance.CELO.shiftedBy(-ERC20_DECIMALS).toFixed(2);
    const cUSDBalance = balance.cUSD.shiftedBy(-ERC20_DECIMALS).toFixed(2);

    setCeloBalance(celoBalance);
    setcUSDBalance(cUSDBalance);
    setloading(false)
  };

  const getGames = async () => {
    setloading(true)
    const _gameLength = await contract.methods.gamesLength().call();
    const _games = [];

    for (let i = 0; i < _gameLength; i++) {
      let _b = new Promise(async (resolve, reject) => {
        let _game = await contract.methods.getGame(i).call();

        resolve({
          index: i,
          owner: _game.owner,
          game_name: _game.name,
          cover_photo: _game.image_link,
          description: _game.description,
          game_file: _game.file_link,
          price: _game.price,
          sales: _game.game_sales,
        });
      });
      _games.push(_b);
    }
    const all_games = await Promise.all(_games);

    setGames(all_games);
    setloading(false)
  };

  useEffect(() => {
    if (contract) return getGames();
  }, [contract]);

  const purchaseGame = async (_price, _index, game_file) => {
    try {
      setloading(true)
      const cUSDContract = new kit.web3.eth.Contract(erc20Abi, ContractAddress);

      await cUSDContract.methods.approve(
        gamifyaddress,
        _price
      ).send({ from: address });

      await contract.methods.buyGame(_index).send({
        from: address
      }).then(() => {
        alert ("Your download will begin shortly from IPFS...")
        window.open(game_file, '_blank') || window.location.replace(game_file);
      });
    
      getBalance();
      getGames();
      setloading(false)
    } catch (error) {
      console.error(error);
      setloading(false)
    }
  };

  // upload game files to IPFS
  const uploadHelper = async (_file) => {
    try {
      const file = await ipfsClient.add(_file);
      console.log(file);
      const path = `https://ipfs.infura.io/ipfs/${file.path}`;
    
      return path;
    } catch (error) {
      console.log("Error uploading file: ", error);
      // return false
      setloading(false)
      throw error;
    }
  };

  // create a game
  const createGame = async (title, price, description, imageFile, gameFile) => {
    setloading(true)
    const image_link = await uploadHelper(imageFile);
    const game_link = await uploadHelper(gameFile);
    const game_price = BigNumber(price).shiftedBy(ERC20_DECIMALS).toString();

    if (!image_link || !game_link) {
      setloading(false)
      return alert("Failed to upload game image or file");
    }

    await contract.methods.addGame(
      title,
      description,
      image_link,
      game_link,
      game_price
    ).send({ from: address });
    

    await getGames();
    getBalance();
    setloading(false)
  };

  return (
    <>
      <Header
        cUSDBalance = {cUSDBalance}
        celoBalance = {celoBalance}
      />
      <Body
        createGame={createGame}
        fileUrl={fileUrl}
        games={games}
        cUSDBalance={cUSDBalance}
        purchaseGame={purchaseGame}
        loading = {loading}
      />
      <Footer />
    </>
  );
}


src/components/Footer.js
import React from 'react'

export default function Footer() {
    return (
        <>

        {/* footer-area */}
<footer>
  <div className="footer-top footer-bg s-footer-bg">
  
    <div className="container">
      <div className="row justify-content-between">
        <div className="col-xl-3 col-lg-4 col-md-6">
          <div className="footer-widget mb-50">
            <div className="footer-logo mb-35">
              {/* <a href="/"><img src="/assets/img/logo/logo.png" alt="" /></a> */}
              <h1>Gamify</h1>
            </div>
            <div className="footer-text">
              <p>The gaming marketplace for you.</p>
             
            </div>
          </div>
        </div>
   
    
      </div>
    </div>
    <div className="footer-fire"><img src="/assets/img/images/footer_fire.png" alt="" /></div>
    <div className="footer-fire footer-fire-right"><img src="/assets/img/images/footer_fire.png" alt="" /></div>
  </div>
  <div className="copyright-wrap s-copyright-wrap">
    <div className="container">
      <div className="text-center">
        <div className="copyright-text">
          <p>Copyright © 2021 <a href="#">Gamify</a> All Rights Reserved.</p>
        </div>
      </div>
    </div>
  </div>
</footer>
{/* footer-area-end */}

            
        </>
    )
}


src/components/Header.js
import React from 'react'

export default function Header(props) {
    return (
        <>
{/* header-area */}
<header className="header-style-four">
  <div className="header-top-area s-header-top-area d-none d-lg-block">
    <div className="container custom-container-two">
      <div className="row align-items-center">
        <div className="col-lg-6 d-none d-lg-block">
          <div className="header-top-offer">
            <p>get instant access to quality games via IPFS</p>
            <span className="coming-time" data-countdown="2021/3/15" />
          </div>
        </div>
        <div className="col-lg-6">
          <div className="header-social">
            <span className="mr-4">Balance : </span>
            <ul>
              <li className="text-center"><h3>{props.cUSDBalance}</h3> cUSD</li>
              <li className="text-center ml-3"><h3>{props.celoBalance}</h3> CELO</li>
              {/* <li><a href="#"><i className="fab fa-twitter" /></a></li>
              <li><a href="#"><i className="fab fa-pinterest-p" /></a></li>
              <li><a href="#"><i className="fab fa-linkedin-in" /></a></li> */}
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div id="sticky-header" className="header-four-wrap">
    <div className="container custom-container-two">
      <div className="row">
        <div className="col-12">
          <div className="main-menu menu-style-two">
            <nav>
              <div className="logo">
                {/* <a href="index-4.html"><img src="/assets/img/logo/logo_two.png" alt="Logo" /></a> */}
                <h1>Gamify</h1>
              </div>
              <div id="mobile-menu" className="navbar-wrap d-none d-lg-flex">
                <ul>
                  <li><a href="#">Home</a>
                    
                  </li>
               </ul>
              </div>
              
            </nav>
          </div>
          <div className="mobile-menu" />
        </div>
        {/* Modal Search */}
        <div className="modal fade" id="search-modal" tabIndex={-1} role="dialog" aria-hidden="true">
          <div className="modal-dialog" role="document">
            <div className="modal-content">
              <form>
                <input type="text" placeholder="Search here..." />
                <button><i className="fa fa-search" /></button>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</header>
{/* header-area-end */}

            
        </>
    )
}


src/components/Body.js
import BigNumber from "bignumber.js";
import React, { useState } from "react";
import { LoadingOverlay, Loader } from 'react-overlay-loader';
 
import 'react-overlay-loader/styles.css';

const ERC20_DECIMALS = 18;

export default function Body(props) {
  const [title, settitle] = useState(null);
  const [description, setdescription] = useState(null);
  const [imageFile, setimageFile] = useState(null);
  const [gameFile, setgameFile] = useState(null);
  const [price, setprice] = useState(0);

  // console.log(props.games)
  return (
    <>
      {/* main-area */}
      <main>
        {/* slider-area */}
        <Loader fullPage loading = {props.loading} />
        <section className="slider-area home-four-slider">
          <div className="slider-active">
            <div className="single-slider slider-bg slider-style-two">
              <div className="container custom-container-two">
                <div className="row">
                  <div className="col-xl-6 col-lg-7 col-md-11">
                    <div className="slider-content">
                      <h6 data-animation="fadeInUp" data-delay=".4s">
                        Gamify
                      </h6>
                      <h2 data-animation="fadeInUp" data-delay=".4s">
                        Sell <span>Blockchain</span> Games
                      </h2>
                      <p data-animation="fadeInUp" data-delay=".6s">
                        Create and sell your favorite games here on the celo
                        blockchain.
                      </p>
                      {/* <a href="#" className="btn btn-style-two" data-animation="fadeInUp" data-delay=".8s">READ MORE</a> */}
                    </div>
                  </div>
                </div>
                <div className="slider-img">
                  <img
                    src="https://themebeyond.com/html/geco/Geco/img/slider/four_slider_img01.png"
                    alt=""
                    data-animation="slideInRightS"
                    data-delay=".8s"
                  />
                </div>
              </div>
            </div>
            <div className="single-slider slider-bg slider-style-two">
              <div className="container custom-container-two">
                <div className="row">
                  <div className="col-xl-6 col-lg-7 col-md-11">
                    <div className="slider-content">
                      <h6 data-animation="fadeInUp" data-delay=".4s">
                        Gamify
                      </h6>
                      <h2 data-animation="fadeInUp" data-delay=".4s">
                        Enjoy <span>Quality</span> Games
                      </h2>
                      <p data-animation="fadeInUp" data-delay=".6s">
                        Get access to tons of games and resources via the celo
                        blockchain.
                      </p>
                      {/* <a href="#" className="btn btn-style-two" data-animation="fadeInUp" data-delay=".8s">READ MORE</a> */}
                    </div>
                  </div>
                </div>
                <div className="slider-img">
                  <img
                    src="https://themebeyond.com/html/geco/Geco/img/slider/third_banner_img.png"
                    alt=""
                    data-animation="slideInRightS"
                    data-delay=".8s"
                  />
                </div>
              </div>
            </div>
          </div>
        </section>
        {/* slider-area-end */}

        {/* shop-area */}
        <section className="shop-area home-four-shop-area pt-115 pb-90">
          <div className="container">
            <div className="row justify-content-center">
              <div className="col-xl-6 col-lg-8">
                <div className="section-title home-four-title text-center mb-35">
                  <h2>
                    gaming product <span>corner</span>
                  </h2>
                  <p>Easily purchase games here for affordable prices</p>
                </div>
              </div>
            </div>
            <div className="row product-active">
              {props.games.map((game, key) => (
                <div className="col-xl-3" key={key}>
                  <div className="shop-item">
                    <div className="product-thumb">
                      <a href="#">
                        <img src={game.cover_photo} alt="" />
                      </a>
                    </div>
                    <div className="product-content">
                      <div className="product-tag">
                        <a href="#">{game.sales} Sales</a>
                      </div>
                      <h4>
                        <a href="#">{game.game_name}</a>
                      </h4>
                      <div className="product-meta">
                        <div className="product-price">
                          <h5>
                            {" "}
                            {new BigNumber(game.price)
                              .shiftedBy(-ERC20_DECIMALS)
                              .toString()}{" "}
                            cUSD
                          </h5>
                        </div>
                        {/* <div className="product-cart-action">
                          <a href="#"><i className="fas fa-shopping-basket" /></a>
                        </div> */}
                        <div className="newsletter-form">
                          <button
                            onClick={() => {
                              props.purchaseGame(game.price, key, game.game_file);
                            }}
                          >
                            Purchase <i class="fas fa-paper-plane"></i>
                          </button>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              ))}
            </div>
          </div>
        </section>
        {/* shop-area-end */}
        {/* <div class="contact-form">
          <form action="#">
            <textarea name="message" id="message" placeholder="Write Message..."></textarea>
            <div class="row">
              <div class="col-md-6">
                <input type="text" placeholder="Your Name">
              </div>
              <div class="col-md-6">
                <input type="email" placeholder="Your Mail">
              </div>
            </div>
            <button>SUBMIT MESSAGE</button>
          </form>
        </div> */}

        <section className="featured-game-area new-released-game-area pt-115 pb-90">
          <div className="container">
            <div className="row justify-content-center">
              <div className="col-lg-6 col-md-8">
                <div className="section-title home-four-title black-title text-center mb-60">
                  <h2>
                    Sell Your <span>GAMES</span>
                  </h2>
                </div>
              </div>
            </div>
          </div>
          <div className="container">
            <div className="justify-content-center align-content-center">
              <div className="col-lg-12 col-md-12 ">
                <div>
                  <div className="contact-form">
                    
                    <form
                      onSubmit={(e) => {
                        e.preventDefault();
                        if (
                          !title ||
                          !price ||
                          !description ||
                          !imageFile ||
                          !gameFile
                        ) {
                          return alert("Please enter all fields");
                        } else {
                          props.createGame(
                            title,
                            price,
                            description,
                            imageFile,
                            gameFile,
                          );
                        }
                        // props.addGane()
                      }}
                    >
                      <div className="row">
                        <div className="col-md-6">
                          <input
                            type="text"
                            required
                            onChange={(e) => {
                              settitle(e.target.value);
                            }}
                            placeholder="Game's Name"
                          />
                        </div>
                        <div className="col-md-6">
                          <input
                            type="number"
                            onChange={(e) => {
                              setprice(Number(e.target.value));
                            }}
                            placeholder="Game's Price in cUSD"
                          />
                        </div>

                        <div className="col-md-12">
                          <textarea
                            name="message"
                            id="message"
                            onChange={(e) => {
                              setdescription(e.target.value);
                            }}
                            placeholder="Game description..."
                          />
                        </div>

                        <div className="col-md-6">
                          <input
                            type="file"
                            onChange={(e) => {
                              setimageFile(e.target.files[0]);
                            }}
                            placeholder="Select game cover photo"
                          />
                          <span className="text-small">Select game cover photo</span>
                        </div>
                        <div className="col-md-6">
                          <input
                            type="file"
                            onChange={(e) => {
                              setgameFile(e.target.files[0]);
                            }}
                            placeholder="Select zipped game file"
                          />
                          <span className="text-small">Select zipped game file</span>
                        </div>
                      </div>
                      <button type="submit" className="mt-5">Click To Create</button>
                    </form>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </section>
      </main>
      {/* main-area-end */}
    </>
  );
}
